### PR TITLE
Handle NaN gracefully in loss computation.

### LINKED ISSUE #370: Robust model fitting (RuntimeError when learning rate is too high)
Thanks for the continued development on GPyTorch. It's fantastic for constructing models but one area that we've recently been struggling with is robustly fitting models, due to numerical issues during optimization. Even a simple ExactGP model errors out during fitting with a RuntimeError if the learning rate is too large.

This makes model fitting a real challenge: too small of a learning rate and it takes forever, too large and it errors. To get around learning rate tuning, we've been using L-BFGS (both from https://github.com/hjmshi/PyTorch-LBFGS and from Scipy), but even this can fail when the L-BFGS takes too large of a step. We want to run this for a large, heterogeneous collection of problems and so really need robust model fitting that doesn't require hyperparameter optimization.

Attached is a notebook with a simple GP example that fails to fit with default L-BFGS setup, or when the Adam lengthscale is too long. It also includes a plot showing the unfortunate hyperparameter sensitivity of Adam.

I've been in touch with @hjmshi about this. Ideally, the L-BFGS would be given a chance to backtrack from the overly large step by getting a NaN for the function return and leaving the model in a useable state. Is there anything that can be done along these lines to make fitting more robust?

[gpytorch_optim_test5.zip](https://github.com/cornellius-gp/gpytorch/files/2578640/gpytorch_optim_test5.zip)

### PR DESCRIPTION
**Pull request by jacobrgardner**
This PR makes the GPyTorch internals handle NaNs that occur due to round-off from terrible hyperparameter values more gracefully. In particular:

- If the preconditioner contains NaNs, we throw a warning and attempt to continue without preconditioning. 
    - In testing these NaNs appear somewhat intermittently such that even if one iteration fails due to it the next will often succeed. In @bletham's notebook this avoids crashing with learning rates well up in to the thousands (when combined with `param_transform=softplus` of course).
- If NaNs are encountered from CG outputs, these are handled more gracefully. 
    - The linear solves already handle these gracefully in the sense that we don't try to do anything more than basic operations, so the inv quad term will already show up as NaN. All that was needed was a check for the tridiagonal matrices used for log det calculations.

In my opinion, this closes #370 in combination with #371.

### CODE REVIEW

**Comment by Balandat** [comment]
great, thanks!
